sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java

```diff
@@ -195,84 +229,166 @@ public void getTempFilePrefix(ProcessContext c) {
-    // loading.
-    PCollection<Void> singleton = p.apply("singleton",
-        Create.of((Void) null).withCoder(VoidCoder.of()));
-    PCollectionTuple partitions =
-        singleton.apply(
-            "WritePartition",
-            ParDo.of(
-                    new WritePartition<>(
-                        singletonTable,
-                        dynamicDestinations,
-                        tempFilePrefix,
-                        resultsView,
-                        multiPartitionsTag,
-                        singlePartitionTag))
-                .withSideInputs(tempFilePrefix, resultsView)
-                .withOutputTags(multiPartitionsTag, TupleTagList.of(singlePartitionTag)));
+    PCollectionTuple partitions;
+    if (triggeringFrequency == null) {
+      // Turn the list of files and record counts in a PCollectionView that can be used as a
+      // side input.
+      PCollectionView<Iterable<WriteBundlesToFiles.Result<DestinationT>>> resultsView =
+          results.apply("ResultsView",
+              View.<WriteBundlesToFiles.Result<DestinationT>>asIterable());
+      // This transform will look at the set of files written for each table, and if any table has
+      // too many files or bytes, will partition that table's files into multiple partitions for
+      // loading.
+      PCollection<Iterable<Result<DestinationT>>> singleton = p.apply("singleton",
+          Create.<Iterable<Result<DestinationT>>>of(ImmutableList.<Result<DestinationT>>of())
+              .withCoder(IterableCoder.of(WriteBundlesToFiles.ResultCoder.of(destinationCoder))));
+      partitions =
+          singleton.apply(
+              "WritePartitionUntriggered",
+              ParDo.of(
+                  new WritePartition<>(
+                      singletonTable,
+                      dynamicDestinations,
+                      tempFilePrefix,
+                      resultsView,
+                      multiPartitionsTag,
+                      singlePartitionTag))
+                  .withSideInputs(tempFilePrefix, resultsView)
+                  .withOutputTags(multiPartitionsTag, TupleTagList.of(singlePartitionTag)));
+    } else {
+      // Apply the user's trigger to ensure we don't generate too many BigQuery load jobs.
+      results  =
+          results.apply(
+              "applyUserTrigger",
+              Window.< WriteBundlesToFiles.Result<DestinationT>>into(new GlobalWindows())
+                  .triggering(Repeatedly.forever(
+                      AfterProcessingTime.pastFirstElementInPane()
+                          .plusDelayOf(triggeringFrequency)))
+                  .discardingFiredPanes());
+
+      // If we have non-default triggered output, the above side-input path is incorrect. Instead
+      // make the result list a main input. Apply a GroupByKey first for determinism.
+      partitions = results
+          .apply("AttachSingletonKey",
```

Comment:
Outside the scope of this PR: we should probably have a transform that converts a PCollection<T> to a `PCollection<Iterable<T>>` like this, by grouping it with a null key (Perhaps something like ReifyContents.of() and ReifyContents.windowed() that will return a `PCollection<KV<BoundedWindow, Iterable<T>>>`). This repeats a bunch of times in the SDK.